### PR TITLE
interfaces/builtin: first cut at repowerd interface

### PR DESCRIPTION
**Pull request by afrantzis**

### CODE REVIEW

File: interfaces/builtin/repowerd.go [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+
+# logind power
+dbus (send)
+  bus=system
+  path=/org/freedesktop/login1
+  interface=org.freedesktop.login1.Manager
+  member={Hibernate,HybridSleep,Inhibit,PowerOff,Suspend}
+  peer=(name=org.freedesktop.login1),
+
+# UPower
```
Comment (morphis) [inline interfaces/builtin/repowerd.go]: This can be replace with the upcoming upower-control interface I am working on.

File: interfaces/builtin/repowerd.go [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+  member=Get{,All}
+  peer=(name=org.freedesktop.UPower),
+dbus (receive)
+  bus=system
+  path=/org/freedesktop/UPower{,/devices/**}
+  interface=org.freedesktop.DBus.Properties
+  member=PropertiesChanged
+  peer=(name=org.freedesktop.UPower),
+
+# Ofono
```
Comment (morphis) [inline interfaces/builtin/repowerd.go]: We have an ofono interface already for this. No need to repeat those bits here. If something is missing in the ofono interface lets add it there.

File: interfaces/builtin/repowerd.go [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+  member=Get{,All}
+  peer=(name=org.ofono),
+dbus (receive)
+  bus=system
+  path=/{,**}
+  interface=org.freedesktop.DBus.Properties
+  member=PropertiesChanged
+  peer=(name=org.ofono),
+
+# Unity.Display
```
Comment (morphis) [inline interfaces/builtin/repowerd.go]: Aren't those Unity.* bits something which should go into a unity interface? unity7 already exists.

File: interfaces/builtin/repowerd.go (hunk 1 above)
Comment (morphis) [inline interfaces/builtin/repowerd.go]: See https://github.com/snapcore/snapd/pull/2498

File: interfaces/builtin/repowerd.go [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+
+# Can communicate with DBus system service
+sendmsg
+sendto
+recvfrom
+recvmsg
+`)
+
+var repowerdPermanentSlotDBus = []byte(`
+<policy user="root">
```
Comment (morphis) [inline interfaces/builtin/repowerd.go]: This this policy directly from upstream or was modified? Please add a note like

File: interfaces/builtin/repowerd.go [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+func (iface *RepowerdInterface) SanitizePlug(plug *interfaces.Plug) error {
+	return nil
+}
+
+func (iface *RepowerdInterface) SanitizeSlot(slot *interfaces.Slot) error {
+	return nil
+}
+
+func (iface *RepowerdInterface) AutoConnect(*interfaces.Plug, *interfaces.Slot) bool {
+	return true
```
Comment (morphis) [inline interfaces/builtin/repowerd.go]: Add comment

// allow what declarations allowed 

above.

File: interfaces/builtin/repowerd.go [hunk 6]
```diff
@@ -0,0 +1,335 @@
...
+# Usage: reserved
+
+# HW alarms
+/dev/alarm rw,
+
+# Suspend
+/sys/power/** rw,
+
+# Backlight
+/sys/**/brightness rw,
```
Comment (morphis) [inline interfaces/builtin/repowerd.go]: These seem to be pretty widespread. Isn't it more a

/sys/class/backlight/{,**} rw,

?

File: interfaces/builtin/repowerd.go (hunk 2 above)
Comment (afrantzis) [inline interfaces/builtin/repowerd.go]: See comment about upower.

File: interfaces/builtin/repowerd.go (hunk 3 above)
Comment (afrantzis) [inline interfaces/builtin/repowerd.go]: These are dbus bits specific to Unity System Compositor (USC). We could add a USC specific interface if that is preferable, but like ofono and upower, I would like snaps not to care about these details (just use a repowerd slot/plug).

File: interfaces/builtin/repowerd.go [hunk 7]
```diff
@@ -0,0 +1,335 @@
...
+  path=/com/canonical/Unity/PowerButton
+  member={Press,Release},
+dbus (send)
+  bus=system
+  interface=com.canonical.Unity.PowerButton
+  path=/com/canonical/Unity/PowerButton
+  member=LongPress,
+`)
+
+var repowerdConnectedPlugAppArmor = []byte(`
```
Comment (afrantzis) [inline interfaces/builtin/repowerd.go]: Thanks, I will look into this.

File: interfaces/builtin/repowerd.go (hunk 4 above)
Comment (afrantzis) [inline interfaces/builtin/repowerd.go]: Directly from upstream, I will add a note.

File: interfaces/builtin/repowerd.go (hunk 6 above)
Comment (afrantzis) [inline interfaces/builtin/repowerd.go]: Unfortunately backlights, especially on phablet devices, are not necessarily under /sys/class/backlight. I will do a survey of our existing devices to see if I can restrict the rule further.

File: interfaces/builtin/repowerd.go (hunk 1 above)
Comment (morphis) [inline interfaces/builtin/repowerd.go]: You shouldn't mix plug and slot here. What clients of repowerd only need is a plug of the repowerd interface. They don't need to use the ofono, .. interfaces as this is only a thing the repowerd snap itself needs to use in order to get access to those services. We're using this scheme all over the place already and hardcoding everything into a single interface just makes maintenance pretty hard.

File: interfaces/builtin/repowerd.go (hunk 3 above)
Comment (morphis) [inline interfaces/builtin/repowerd.go]: As said above, we're talking about the slot here and not the plug. You're right clients should be able to use just the repowerd plug to talk with repowerd and thats it. But I guess today no client talks with ofono directly and if it does its an implementation detail of that snap and doesn't relate to the use of repowerd.

File: interfaces/builtin/repowerd.go (hunk 1 above)
Comment (afrantzis) [inline interfaces/builtin/repowerd.go]: Removed custom rules in favor of using upower-observe.

File: interfaces/builtin/repowerd.go (hunk 2 above)
Comment (afrantzis) [inline interfaces/builtin/repowerd.go]: Removed custom rules in favor of using ofono plug.

File: interfaces/builtin/repowerd.go (hunk 7 above)
Comment (afrantzis) [inline interfaces/builtin/repowerd.go]: Done.

File: interfaces/builtin/repowerd.go (hunk 4 above)
Comment (afrantzis) [inline interfaces/builtin/repowerd.go]: Added xml comment.

File: interfaces/builtin/repowerd.go (hunk 5 above)
Comment (afrantzis) [inline interfaces/builtin/repowerd.go]: Added comment.

File: interfaces/builtin/repowerd.go (hunk 6 above)
Comment (afrantzis) [inline interfaces/builtin/repowerd.go]: I did the survey, but unfortunately I am not able to restrict the rule further. Although repowerd reads backlight info from subdirs of /sys/class/backlight/ (and /sys/class/leds/lcd-backlight on some devices) these subdirs are actually links to arbitrary paths within /sys/... , depending on device, graphics card etc. Using a /sys/class/backlight/... rule doesn't work since apparmor also checks permissions for the linked directories, not just the links themselves.

**Comment by afrantzis** [comment]
ping

File: interfaces/builtin/repowerd.go (hunk 6 above)
Comment (niemeyer) [inline interfaces/builtin/repowerd.go]: That's awkward.. we can't allow arbitrary access to anything named "brightness" under sys, no matter what it is under.

**Comment by niemeyer** [comment]
@jdstrand can you please have a pass here when you have a moment?

File: interfaces/builtin/repowerd.go (hunk 6 above)
Comment (niemeyer) [inline interfaces/builtin/repowerd.go]: Well, we always can, but ideally we wouldn't. Given that the slot availability will be pretty restricted nevertheless, since it's a permanent one, we might allow it in. @jdstrand: any ideas?

**Comment by niemeyer** [comment]
@afrantzis Can you please make sure the PR passes the integration tests?

**Comment by niemeyer** [comment]
@jdstrand @afrantzis pings

File: interfaces/builtin/repowerd.go (hunk 6 above)
Comment (jdstrand) [inline interfaces/builtin/repowerd.go]: This is a rough spot in our implementation and exists elsewhere (eg udev with /sys and /run/udev). For a slot implementation, it is not ideal to have a glob rule, but it also isn't the worst thing, so I'm ok with this type of rule if it can be scoped some. `/sys/**/type r,` (below) is pretty icky for example. Granted, it is read only, but there are a *lot* of files named 'type' in /sys (kernel, usb, firmware, etc).

Would this work:
```
# Allow adjusting backlights
# TODO: fine-tune this to resolve symlinks to use fine-grained rules
￼/sys/devices/{platform,pci*}/**/brightness rw,
/sys/devices/{platform,pci*}/**/max_brightness r,
/sys/devices/{platform,pci*}/**/type r,
```

One way to avoid this would be to declare to snapd what brightness files we are interested in (eg, /sys/class/backlight/* and /sys/class/leds/lcd-backlight/*), then have snapd resolve the symlinks to absolute paths, then create a sorted list of apparmor rules to put in the policy. This would work fine for backlights because those won't be hotpluggable. It also is cheap for snapd to do as part of its state operations and because the list is sorted, the policy shouldn't ever change so there is no apparmor parser overhead with this approach.

File: interfaces/builtin/repowerd.go [hunk 8]
```diff
@@ -0,0 +1,307 @@
...
+/sys/**/type r,
+
+# DBus
+#include <abstractions/dbus-strict>
+
+dbus (send)
+  bus=system
+  path=/org/freedesktop/DBus
+  interface=org.freedesktop.DBus
+  member={GetId,RequestName,ReleaseName,GetConnectionUnixProcessID}
```
Comment (jdstrand) [inline interfaces/builtin/repowerd.go]: What about GetConnectionUnixUser?

File: interfaces/builtin/repowerd.go [hunk 9]
```diff
@@ -0,0 +1,307 @@
...
+  peer=(name=org.freedesktop.DBus),
```
Comment (jdstrand) [inline interfaces/builtin/repowerd.go]: Shouldn't this be: `peer=(name=org.freedesktop.DBus, label=unconfined)`?

File: interfaces/builtin/repowerd.go [hunk 10]
```diff
@@ -0,0 +1,307 @@
...
+
+dbus (receive, send)
+  bus=system
+  interface=org.freedesktop.DBus.{Properties,Introspectable},
```
Comment (jdstrand) [inline interfaces/builtin/repowerd.go]: This is too lenient and allows getting properties of all services on the system bus. What denial did you see that made you put this in the policy?

File: interfaces/builtin/repowerd.go [hunk 11]
```diff
@@ -0,0 +1,307 @@
...
+
+dbus (bind)
+  bus=system
+  name=com.canonical.powerd,
+
+dbus (bind)
+  bus=system
+  name=com.canonical.Unity.Screen,
```
Comment (jdstrand) [inline interfaces/builtin/repowerd.go]: repowerd binds to two different well-known names? Should this be two different interfaces and repowerd slots both? Why is the well-known name com.canonical.powerd instead of com.canonical.repowerd? Should the interface be name 'powerd' instead?

File: interfaces/builtin/repowerd.go [hunk 12]
```diff
@@ -0,0 +1,307 @@
...
+  bus=system
+  interface=com.canonical.Unity.PowerButton
+  path=/com/canonical/Unity/PowerButton
+  member={Press,Release},
+dbus (send)
+  bus=system
+  interface=com.canonical.Unity.PowerButton
+  path=/com/canonical/Unity/PowerButton
+  member=LongPress,
+`)
```
Comment (jdstrand) [inline interfaces/builtin/repowerd.go]: Also, dbus rule style in snapd uses 4 spaces on the continuation lines instead of 2.

File: interfaces/builtin/repowerd.go [hunk 13]
```diff
@@ -0,0 +1,307 @@
...
+  bus=system
+  name=com.canonical.Unity.Screen,
+
+dbus (receive, send)
+  bus=system
+  interface=com.canonical.powerd,
+
+dbus (receive, send)
+  bus=system
+  interface=com.canonical.Unity.Screen,
```
Comment (jdstrand) [inline interfaces/builtin/repowerd.go]: These are too lenient. They should be receive rules, have a 'path' element and also use 'peer=(label=unconfined)'. Eg:
```
# Allow unconfined processes to communicate with repowerd
dbus (receive)
    bus=system
    path=/com/canonical/powerd{,/**}
    interface=com.canonical.powerd{,.*}
    peer=(label=unconfined),

dbus (receive)
    bus=system
    path=/com/canonical/Unity/Screen{,/**}
    interface=com.canonical.Unity.Screen{,.*}
    peer=(label=unconfined),
```

File: interfaces/builtin/repowerd.go [hunk 14]
```diff
@@ -0,0 +1,307 @@
...
+  path=/org/freedesktop/login1
+  interface=org.freedesktop.login1.Manager
+  member={SessionAdded,SessionRemoved}
+  peer=(name=org.freedesktop.login1),
+dbus (receive)
+  bus=system
+  path=/org/freedesktop/login1{,/seat/**,/session/**}
+  interface=org.freedesktop.DBus.Properties
+  member=PropertiesChanged
+  peer=(name=org.freedesktop.login1),
```
Comment (jdstrand) [inline interfaces/builtin/repowerd.go]: All of these should use:
```
peer=(name=org.freedesktop.login1, label=unconfined)
```

but beyond that, this seems unrelated to repowerd. Why is repowerd interested in all of this and what is it doing with the information? With the exception of GetSessionByPID, this is all very similar to the PermanentSlotAppArmor policy in upower-observe....

File: interfaces/builtin/repowerd.go [hunk 15]
```diff
@@ -0,0 +1,307 @@
...
+
+# logind power
+dbus (send)
+  bus=system
+  path=/org/freedesktop/login1
+  interface=org.freedesktop.login1.Manager
+  member={Hibernate,HybridSleep,Inhibit,PowerOff,Suspend}
+  peer=(name=org.freedesktop.login1),
```
Comment (jdstrand) [inline interfaces/builtin/repowerd.go]: This should also use `label=unconfined`

File: interfaces/builtin/repowerd.go [hunk 16]
```diff
@@ -0,0 +1,307 @@
...
+
+# Unity.Display
+dbus (send)
+  bus=system
+  interface=com.canonical.Unity.Display
+  path=/com/canonical/Unity/Display
+  member={TurnOn,TurnOff}
+  peer=(name=com.canonical.Unity.Display),
```
Comment (jdstrand) [inline interfaces/builtin/repowerd.go]: Is this part of the unity8 interface? If not, what is providing this service? If so, then the unity8 interface or some other unity8 interface should have ConnectedSlot and ConnectedPlug policy such that repowerd may `plugs: [ unity8 ]`. As written, it assumes that Unity.Display is around, but it doesn't exist in the core snap and therefore you need to connect this snap to the snap that provides this service. More concretely, we need a `peer=(label=...)` match but we don't have anything to set the label to.

File: interfaces/builtin/repowerd.go [hunk 17]
```diff
@@ -0,0 +1,307 @@
...
+dbus (receive)
+  bus=system
+  interface=com.canonical.Unity.PowerButton
+  path=/com/canonical/Unity/PowerButton
+  member={Press,Release},
+dbus (send)
+  bus=system
+  interface=com.canonical.Unity.PowerButton
+  path=/com/canonical/Unity/PowerButton
+  member=LongPress,
```
Comment (jdstrand) [inline interfaces/builtin/repowerd.go]: These have the same issue as the rule before them.

File: interfaces/builtin/repowerd.go [hunk 18]
```diff
@@ -0,0 +1,307 @@
...
+`)
+
+var repowerdConnectedPlugAppArmor = []byte(`
+# Description: Allow using repowerd service. Reserved because this gives
+#  privileged access to the repowerd service.
+# Usage: reserved
```
Comment (jdstrand) [inline interfaces/builtin/repowerd.go]: Please remove 'Usage' and have only one space before 'privileged access to the repowerd service'

File: interfaces/builtin/repowerd.go [hunk 19]
```diff
@@ -0,0 +1,307 @@
...
+dbus (receive, send)
+  bus=system
+  path=/com/canonical/powerd
+  peer=(label=###SLOT_SECURITY_TAGS###),
+
+dbus (receive, send)
+  bus=system
+  path=/com/canonical/Unity/Screen
+  peer=(label=###SLOT_SECURITY_TAGS###),
+`)
```
Comment (jdstrand) [inline interfaces/builtin/repowerd.go]: You also need ConnectedSlotAppArmor policy that use ###PLUG_SECURITY_TAGS###.

File: interfaces/builtin/repowerd.go [hunk 20]
```diff
@@ -0,0 +1,307 @@
...
+  bus=system
+  path=/com/canonical/Unity/Screen
+  peer=(label=unconfined),
+`)
+
+var repowerdPermanentSlotSecComp = []byte(`
+# Description: Allow operating as the repowerd service. Reserved because this
+# gives
+#  privileged access to the system.
+# Usage: reserved
```
Comment (jdstrand) [inline interfaces/builtin/repowerd.go]: Please remove 'Usage' and cleanup the comment newline.

File: interfaces/builtin/repowerd.go [hunk 21]
```diff
@@ -0,0 +1,307 @@
...
+sendmsg
+sendto
+recvfrom
+recvmsg
+`)
+
+var repowerdConnectedPlugSecComp = []byte(`
+# Description: Allow using repowerd service. Reserved because this gives
+#  privileged access to the repowerd service.
+# Usage: reserved
```
Comment (jdstrand) [inline interfaces/builtin/repowerd.go]: Please remove 'Usage' and use only one space on the line before.

File: interfaces/builtin/repowerd.go [hunk 22]
```diff
@@ -0,0 +1,307 @@
...
+
+import (
+	"bytes"
+	"fmt"
+
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/release"
+)
+
+var repowerdPermanentSlotAppArmor = []byte(`
```
Comment (jdstrand) [inline interfaces/builtin/repowerd.go]: All of the `var` variables that have security policy should use `const` instead. See upower-observe.go as an example.

File: interfaces/builtin/repowerd.go [hunk 23]
```diff
@@ -0,0 +1,307 @@
...
+  <allow own="com.canonical.Unity.Screen"/>
+  <allow send_destination="com.canonical.Unity.Screen"
+         send_interface="com.canonical.Unity.Screen"/>
+
+  <allow own="com.canonical.powerd"/>
+  <allow send_destination="com.canonical.powerd"
+         send_interface="com.canonical.powerd"/>
+</policy>
+
+<policy context="default">
```
Comment (jdstrand) [inline interfaces/builtin/repowerd.go]: Please add:
```
<deny own="com.canonical.Unity.Screen"/>
<deny own="com.canonical.powerd"/>
```

File: interfaces/builtin/repowerd.go [hunk 24]
```diff
@@ -0,0 +1,307 @@
...
+         send_interface="com.canonical.powerd"
+         send_type="method_call" send_member="userAutobrightnessEnable" />
+</policy>
+
+<!-- Allow any user at console (adb shell) to send commands this is
+     needed for unlocking the screen in tests (LP: 1298869) -->
+<policy at_console="true">
+  <allow send_destination="com.canonical.Unity.Screen"/>
+  <allow send_destination="com.canonical.powerd"/>
+</policy>
```
Comment (jdstrand) [inline interfaces/builtin/repowerd.go]: I don't think this phablet workaround is what we want here. Touch devices had many assumptions that snapd does not. Please remove this and whenever polkit is implemented, add appropriate polkit policy for devices where this makes sense. cc @vosst

File: interfaces/builtin/repowerd.go [hunk 25]
```diff
@@ -0,0 +1,307 @@
...
+		return repowerdPermanentSlotSecComp, nil
+	case interfaces.SecurityDBus:
+		return repowerdPermanentSlotDBus, nil
+	}
+	return nil, nil
+}
+
+func (iface *RepowerdInterface) ConnectedSlotSnippet(plug *interfaces.Plug, slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	return nil, nil
+}
```
Comment (jdstrand) [inline interfaces/builtin/repowerd.go]: See above, this should be implemented.

**Comment by niemeyer** [comment]
Closing this for the time being as there has been two weeks without feedback.

We're still happy to collaborate and integrate it once you have a moment to work on it. Just let us know on your progress and we'll reopen it.

**Comment by afrantzis** [comment]
> We're still happy to collaborate and integrate it once you have a moment to work on it.
> Just let us know on your progress and we'll reopen it.

Sorry for the silence, I have been busy with a new repowerd release. The new release requires a few additions to the interface. I will reopen when the changes are integrated.

**Comment by niemeyer** [comment]
No worries. Please let us know and we'll jump back on it.
